File: 139-word-break/139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        arr = [False for _ in range(len(s)+1)]
        arr[0]=True
        for i in range(len(s)):
            for j in range(i+1):
                if s[j:i+1] in wordDict:
                    arr[i+1] = arr[i+1] or arr[j]
        return arr[-1]

File: 139-word-break/139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int size = s.size();
        unordered_set<string> wordset( wordDict.begin(), wordDict.end());
        vector<bool>dp(size+1, false);
        dp[0]= true;
        for(int i = 0 ; i < size; i++){
            for(int j = 0; j <= i; j++){
                if(wordset.find(s.substr(j,i+1-j)) != wordset.end()){
                    dp[i+1] = dp[i+1] || dp[j];
                    if(dp[i+1])
                        break;
                } 
            }
        }
        return dp[size];
    }
};